### DISCUSSION TITLE
parse( ptr, len ) ?!?

### DISCUSSION THREAD

**Discussion by boris-999** [nlohmann/json#3096] [Q&A] [answered]
i have a uint8_t*  (pointing at UTF-8)  and a size_t len.   how do i pass that to  parse() ?     i DONT have a null terminated string.    to save you some time, im going to preempt some reply about iterators; if so, please explain the syntax enough to actually use them, the docs have been overly complicated on the matter, and the examples seem to be lacking.

**Comment by gregmarr** [comment, marked as answer]
Have you tried `auto j = json::parse(pointer, pointer + len);`?

**Comment by boris-999** [comment]
thanks.  that did it. :)    feel free to ignore the following, as problem solved... bbut.. if you can direct me to why this is in the docs that would help me in future.  why did i fail to find this in docs?    
also, somewhere i read about (ptr,ptr+len) being deprecated, not sure where.. was in one of the 27 web pages i had open trying to find this answer. (vaguely recall it might have been in a recent release notes on github)

**Reply by nlohmann** [reply]
The parse function is documented here: https://json.nlohmann.me/api/basic_json/parse/. The deprecation is for `from_cbor` etc. where a *pair* of `ptr` and `len` is passed as `{ptr, len}`. That is:

- new: `basic_json::from_cbor(ptr, ptr+len)`
- deprecated: `basic_json::from_cbor({ptr, len})`

I am currently streamlining the documentation. Who would have known too much documentation can be an issue ;)

**Reply by boris-999** [reply]
i shall try to explain my confusion:  the examples on that page don't show an example for using 2 iterators.   template parameters paragraph  say `IteratorType: a compatible iterator type`, well, i wasn't expecting an invalid one!  the comment is kinda useless.   maybe the idea that a ptr to the data _is_ magically one of these valid IteratorTypes  might be obvious to you, it's far from obvious to me.

I'm not sure why there are many examples of null-terminted inputs and so few of non-null terminated inputs; in my experience the acquisition of JSON text is rarely null terminated.. either by file read or network packet.

**Reply by nlohmann** [reply]
The page states

> The value_type of the iterator must be a integral type with size of 1, 2 or 4 bytes, which will be interpreted respectively as UTF-8, UTF-16 and UTF-32.

which is true in your case. But you're right. I shall add some examples for iterator ranges.

**Reply by nlohmann** [reply]
I added a PR: https://github.com/nlohmann/json/pull/3100 - feedback welcome!
